### PR TITLE
Initial version for planning-poker slack app

### PR DESCRIPTION
**Pull request by satpal82bhandari**
# MAJOR
# CHANGELOG

This is the initial version of the planning-poker app.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+var express = require('express');
```
Comment (psbanka) [inline app.js]: There is no reason to use `var` any longer, assuming you can use at least version `5.0` of node.

File: auth.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+var https = require('https');
+var path = require('path');
+var fs = require('fs');
+var auth = {};
+
+//To get the oauth token from slack server.
+auth.getToken = function (req, res, next) {
+
+ var authCode =  req.query.code;
+ var slackKey = JSON.parse(fs.readFileSync(path.join(__dirname + '/config/slackAppKey.json'), 'utf8'));
```
Comment (psbanka) [inline auth.js]: Would like to see a `README.md` file for how to set this system up? This should include documentation on the process of registering the application with Slack, setting up the Heroku server, getting the slack keys, etc.

File: auth.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+ console.log(slackKey);
+ var encodedUrlRedirect = encodeURIComponent('https://orchestraion-india.slack.com/messages');
+ var extServerOptions = {
+  hostname: 'slack.com',
+  path: '/api/oauth.access?client_id='+slackKey.client_id+'&client_secret='+slackKey.client_secret+'&code='+authCode,
```
Comment (psbanka) [inline auth.js]: Better to use `querystring()` (https://nodejs.org/api/querystring.html):

```
querystring.stringify({ foo: 'bar', baz: ['qux', 'quux'], corge: '' })
// returns 'foo=bar&baz=qux&baz=quux&corge='
```

File: config/data.json [hunk 4]
```diff
@@ -0,0 +1,120 @@
+{
```
Comment (psbanka) [inline config/data.json:1]: We need to get proper linting set up in this project. Tabs are unacceptable in source files.

File: config/data.json [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	"attachments": [{
+			"text": "Please give your vote",
+			"color": "#3AA3E3",
+			"attachment_type": "default",
+			"callback_id": "planning-poker",
+			"actions": [{
+				"name": "one",
+				"text": "1",
```
Comment (psbanka) [inline config/data.json:9]: An awful lot of repetition in this file. Can't this be factored out?

File: config/slackAppKey.json [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (psbanka) [inline config/slackAppKey.json]: I have to assume that you did not mean to check this file in? Seems like `config/slackAppKey.json` should be added to your `.gitignore`

File: model/userrating.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
+function UserRating(userid,userName,rating) {
+  this.userid = userid;
+  this.userName = userName;
+  this.rating = (typeof rating !== 'undefined') ? parseInt(rating) :1;
```
Comment (psbanka) [inline model/userrating.js]: no reason to use `typeof` any longer with modern javascript. Just do:
`this.rating = rating !== undefined ? parseInt(rating) : 1`

File: model/userrating.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+function UserRating(userid,userName,rating) {
```
Comment (psbanka) [inline model/userrating.js]: jsdocs. Please follow the jsdocs standards we use for all projects. Each function must have a full jsdoc docstring with `@param` for each parameter and `@returns` if it returns. Also, this module should be called `user-rating.js` and it should have a docstring at the top of it explaining what it is for.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (psbanka) [inline package.json:1]: Please see https://github.com/ciena-blueplanet/pr-bumper/blob/master/package.json for the proper way to set up this file

File: pokerbot.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+var pokerbot = {};
+var ratingModel = {};    // It conatins the rating by each user. It is a map having key as unique Jira ID and value has an array comtaining ratingModel objects.
+
+pokerbot.root = function (req, res, next) {
+ var requestBodyTextArray =  req.body.text.split(" ");
+ var option =  requestBodyTextArray[0] ? requestBodyTextArray[0] : undefined;
+ var jiraId = requestBodyTextArray[1] ? requestBodyTextArray[1] : undefined;
+
+ //Bad command syntex.
```
Comment (psbanka) [inline pokerbot.js]: Spelling error: `syntax`

File: pokerbot.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
+var app = require('./app');
+var fs = require('fs');
+var path = require('path');
+var util = require('./util');
+var UserRating = require('./model/userrating');
+
+var pokerbot = {};
+var ratingModel = {};    // It conatins the rating by each user. It is a map having key as unique Jira ID and value has an array comtaining ratingModel objects.
+
+pokerbot.root = function (req, res, next) {
```
Comment (psbanka) [inline pokerbot.js:12]: jsdocs

File: pokerbot.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    console.log(ratingModel);
+
+    var response_planning_complete = {
+     response_type: "in_channel",
+     text: "Planning for this JIRA ID is complete. Thanks for Playing."
+    }
+    // write the complete logic for average calculation.
+    var userRatingArray = ratingModel[jiraId];
+    console.log(userRatingArray);
+    //var sortedUserRatingArray = util.sortArrayBasedOnObjectProperty(userRatingArray,'rating');
```
Comment (psbanka) [inline pokerbot.js]: Remove commented code

File: pokerbot.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+ //Starting the duplicate jira.
+ if(option=='start' && (ratingModel.hasOwnProperty(jiraId))){
+  console.log(ratingModel);
+  var  response_duplicate_jira = {
+    text: "Planning for this Jira : "+jiraId+ " is already in progress"
+   }
+   return res.status(200).json(response_duplicate_jira);
+ }
+
+ //Star the Poker game.
```
Comment (psbanka) [inline pokerbot.js]: Typo in comment

File: pokerbot.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+ if(option=='start' && (!ratingModel.hasOwnProperty(jiraId))){
+  console.log(ratingModel);
+  ratingModel[jiraId] = new Array();
+  var attachments = JSON.parse(fs.readFileSync(path.join(__dirname + '/config/data.json'), 'utf8'));
+  var  response = {
+    response_type: "in_channel",
```
Comment (psbanka) [inline pokerbot.js]: Magic string. Pull things like this out to the top and define them as constants, eg.

```
const IN_CHANNEL_RESPONSE = 'in_channel'
...
const response = {
  'response_type': IN_CHANNEL_RESPONSE
...
```

File: pokerbot.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+    text: "Please give your poker vote for "+jiraId,
+    attachments : attachments.attachments
+   }
+   return res.status(200).json(response);
+ }
+
+}
+
+pokerbot.vote = function (req, res, next) {
```
Comment (psbanka) [inline pokerbot.js:79]: jsdoc

File: util.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
+var app = require('./app');
+var https = require('https');
+var util = {};
+
+// This is utilty method for getting channel info from slack server.
+util.getUserCountInChannel = function (token,channelId) {
```
Comment (psbanka) [inline util.js]: Proper jsdocs, rather than just a comment

File: util.js [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  });
+ });
+ req.end();
+ req.on('error', (e) => {
+  console.error(e);
+  return JSON.parse(d);
+ });
+}
+
+util.sortArrayBasedOnObjectProperty = function(items,prop){
```
Comment (psbanka) [inline util.js]: jsdocs

**Comment by sglanzer-fire** [comment]
Seems like this belongs in `ciena-blueplanet` not `ciena-frost`

File: config/data.json (hunk 4 above)
Comment (satpal82bhandari) [inline config/data.json:1]: Done.

File: config/slackAppKey.json (hunk 6 above)
Comment (satpal82bhandari) [inline config/slackAppKey.json]: done with blank keys.

File: auth.js (hunk 3 above)
Comment (satpal82bhandari) [inline auth.js]: done

File: model/userrating.js (hunk 7 above)
Comment (satpal82bhandari) [inline model/userrating.js]: done

File: package.json (hunk 9 above)
Comment (satpal82bhandari) [inline package.json:1]: done

File: pokerbot.js (hunk 13 above)
Comment (satpal82bhandari) [inline pokerbot.js]: done

File: pokerbot.js (hunk 14 above)
Comment (satpal82bhandari) [inline pokerbot.js]: done

File: pokerbot.js (hunk 10 above)
Comment (satpal82bhandari) [inline pokerbot.js]: done

File: package.json [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+  "name": "planning-poker-for-slack",
+  "version": "0.1.0",
+  "description": "Planning poker for spint planning on slack",
+  "main": "app.js",
+  "scripts": {
+    "lint": "eslint  *.js model config tests",
+    "test": "npm run-script lint && npm run-script intergartion-test",
```
Comment (psbanka) [inline package.json:8]: `integration`

File: config/data.json (hunk 5 above)
Comment (psbanka) [inline config/data.json:9]: @satpal82bhandari did this ever get fixed?

Comment (satpal82bhandari) [inline config/data.json:9]: done.

File: package.json (hunk 18 above)
Comment (satpal82bhandari) [inline package.json:8]: done

File: auth.js (hunk 2 above)
Comment (satpal82bhandari) [inline auth.js]: done

File: pokerbot.js (hunk 11 above)
Comment (satpal82bhandari) [inline pokerbot.js:12]: done

**Pull request merged**